### PR TITLE
Add plot filling

### PR DESCRIPTION
**Pull request by Wertzui123**
## Introduction
This feature was suggested in #302.
~~**This is not working yet! It's a draft.**~~

### Relevant issues
This feature was suggested in #302.

### API changes
Added the MyPlot::fillPlot() method.

### Behavioural changes
I don't think so.

## Backwards compatibility
Should be backwards compatible.

## Follow-up
Requires translations:

Look at the commit

## Tests
This did not work for me on a test server when I opened the pr.
The plot was not filling with stone when I typed /p fill stone. I would be happy if anyone figured out the problem, otherwise, I'll look at it myself in the next days.

### CODE REVIEW

File: plugin.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name: MyPlot
 main: MyPlot\MyPlot
-version: 1.9.0
+version: 1.10
```
Comment (jasonwynn10) [inline plugin.yml:3]: Please add a `.0` on the end of the version

File: src/MyPlot/MyPlot.php [hunk 2]
```diff
@@ -615,7 +617,7 @@ public function clearPlot(Plot $plot, int $maxBlocksPerTick = 256) : bool {
 		}
 		$maxBlocksPerTick = $ev->getMaxBlocksPerTick();
 		foreach($this->getServer()->getLevelByName($plot->levelName)->getEntities() as $entity) {
-			if($this->getPlotBB($plot)->isVectorInXZ($entity)) {
+			if($this->getPlotBB($plot)->isVectorInXZ($entity) && $entity->getY() <= $this->getLevelSettings($plot->levelName)->groundHeight) {
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:620]: This change is unrelated to this PR

File: src/MyPlot/MyPlot.php [hunk 3]
```diff
@@ -686,6 +688,77 @@ public function clearPlot(Plot $plot, int $maxBlocksPerTick = 256) : bool {
...
+            $selection->setPosition(2, new Vector3($plotBeginPos->x + $plotSize, $plotLevel->groundHeight, $plotBeginPos->z + $plotSize));
+            $cuboid = Cuboid::fromSelection($selection);
+            //$cuboid = $cuboid->async();
+            $cuboid->set($plotBeginPos->level, $plotFillBlock, function (float $time, int $changed) use ($plugin) : void {
+                $plugin->getLogger()->debug('Set ' . number_format($changed) . ' blocks in ' . number_format($time, 10) . 's');
+            });
+            $styler->removeSelection(99998);
+            // Bottom of world
+            $selection = $styler->getSelection(99998);
+            $plotBeginPos->y = 0;
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:744]: Should fill change the bedrock layer too?

File: src/MyPlot/MyPlot.php [hunk 4]
```diff
@@ -1016,9 +1089,9 @@ private function teleportMiddle(Player $player, Plot $plot) : bool {
 	/* -------------------------- Non-API part -------------------------- */
 	public function onLoad() : void {
 		if (!\class_exists(SpoonDetector::class)) {
-			$this->getLogger()->critical("SpoonDetector Virion not found! Please re-download MyPlot from Poggit.");
-			return;
-		}
+            $this->getLogger()->critical("SpoonDetector Virion not found! Please re-download MyPlot from Poggit.");
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:1092]: Don't change between tabs and spaces please

File: src/MyPlot/MyPlot.php [hunk 5]
```diff
@@ -1114,23 +1187,24 @@ public function onLoad() : void {
 		$this->getServer()->getCommandMap()->register("myplot", new Commands($this));
 	}
 
-	public function onEnable() : void {
+	public function onEnable() : void
+	{
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:1191]: formatting

File: src/MyPlot/MyPlot.php [hunk 6]
```diff
@@ -1114,23 +1187,24 @@ public function onLoad() : void {
...
 		if (!\class_exists(SpoonDetector::class)) {
 			$this->getServer()->getPluginManager()->disablePlugin($this);
 			return;
 		}
 		SpoonDetector::printSpoon($this, "spoon.txt");
-		if($this->isDisabled()) {
+		if ($this->isDisabled()) {
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:1197]: formatting

File: src/MyPlot/MyPlot.php [hunk 7]
```diff
@@ -1114,23 +1187,24 @@ public function onLoad() : void {
...
 			return;
 		}
 		$this->getLogger()->debug(TF::BOLD . "Loading Events");
 		$eventListener = new EventListener($this);
 		$this->getServer()->getPluginManager()->registerEvents($eventListener, $this);
 		$this->getLogger()->debug(TF::BOLD . "Registering Loaded Levels");
-		foreach($this->getServer()->getLevels() as $level) {
+		foreach ($this->getServer()->getLevels() as $level) {
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:1204]: formatting

File: src/MyPlot/MyPlot.php [hunk 8]
```diff
@@ -1114,23 +1187,24 @@ public function onLoad() : void {
...
 			$eventListener->onLevelLoad(new LevelLoadEvent($level));
 		}
-		$this->getLogger()->debug(TF::BOLD.TF::GREEN."Enabled!");
+		$this->getLogger()->debug(TF::BOLD . TF::GREEN . "Enabled!");
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:1207]: formatting

File: src/MyPlot/MyPlot.php [hunk 9]
```diff
@@ -686,6 +688,77 @@ public function clearPlot(Plot $plot, int $maxBlocksPerTick = 256) : bool {
 		return true;
 	}
 
+    /**
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:691]: Please use tabs in formatting

File: src/MyPlot/forms/subforms/FillForm.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
+<?php
```
Comment (jasonwynn10) [inline src/MyPlot/forms/subforms/FillForm.php:1]: Please use tabs in formatting

File: src/MyPlot/forms/subforms/FillForm.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    public function __construct(Player $player) {
+        $plugin = MyPlot::getInstance();
+
+        if(!isset($this->plot))
+            $this->plot = $plugin->getPlotByPosition($player);
+
+        parent::__construct(
+            TextFormat::BLACK.$plugin->getLanguage()->translateString("form.header", [$plugin->getLanguage()->get("fill.form")]),
+            [
+                new Input(
```
Comment (jasonwynn10) [inline src/MyPlot/forms/subforms/FillForm.php:22]: Possibly add a second option as a dropdown menu with block string names and check which is value was entered in the form response?
I can make this change after the PR is merged if necessary

File: src/MyPlot/task/FillPlotTask.php [hunk 12]
```diff
@@ -0,0 +1,111 @@
+<?php
```
Comment (jasonwynn10) [inline src/MyPlot/task/FillPlotTask.php:1]: Please use tabs in formatting

File: resources/deu.ini [hunk 13]
```diff
@@ -271,7 +278,7 @@ clone.name=clone
 clone.desc=Klont ein Grundstück an eine andere Position
 clone.usage=/p clone <id: string> [welt: string]
 clone.wrongid=Die Grundstück ID sollte in diesem Format sein: X;Z
-clone.nomoney=Du besitzt nicht genug Geld um dieses Plot zu klonen
+clone.nomoney=Du besitzt nicht genug Geld um dieses Grundstück zu klonen
```
Comment (jasonwynn10) [inline resources/deu.ini:281]: This change is outside the scope of this PR

File: resources/deu.ini [hunk 14]
```diff
@@ -282,10 +289,10 @@ clone.formworld=Grundstückswelt
 
 ; the /p kick Subcommand
 kick.name=kick
-kick.desc=Kicke einen Spieler von deinem Plot
+kick.desc=Kicke einen Spieler von deinem Grundstück
```
Comment (jasonwynn10) [inline resources/deu.ini:292]: This change is outside the scope of this PR

File: resources/deu.ini [hunk 15]
```diff
@@ -282,10 +289,10 @@ clone.formworld=Grundstückswelt
...
 kick.usage=/p kick <spieler: target>
 kick.noPlayer=Der angegebene Spieler konnte nicht gefunden werden
-kick.notInPlot=Der angegebene Spieler befindet sich nicht auf deinem Plot
+kick.notInPlot=Der angegebene Spieler befindet sich nicht auf deinem Grundstück
```
Comment (jasonwynn10) [inline resources/deu.ini:295]: This change is outside the scope of this PR

File: resources/deu.ini [hunk 16]
```diff
@@ -5,7 +5,7 @@
 ; Popup when entering a plot
 popup=Du hast das Grundstück {%0} §rbetreten
 popup.owner=Dieses Grundstück gehört {%0}
-popup.forsale=Dieses Plot wurde von {%0} zum Verkauf erklärt. Mit /p buy kannst du dir dieses Grundstück für {%1}$ kaufen.
+popup.forsale=Dieses Grundstück wurde von {%0} zum Verkauf erklärt. Mit /p buy kannst du dir dieses Grundstück für {%1}$ kaufen.
```
Comment (jasonwynn10) [inline resources/deu.ini:8]: This change is outside the scope of this PR

File: src/MyPlot/subcommand/FillSubcommand.php [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	/**
+	 * @param Player $sender
+	 * @param string[] $args
+	 *
+	 * @return bool
+	 */
+	public function execute(CommandSender $sender, array $args) : bool {
+        if(empty($args)) {
```
Comment (jasonwynn10) [inline src/MyPlot/subcommand/FillSubcommand.php:32]: formatting

File: src/MyPlot/subcommand/FillSubcommand.php [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+		if($plot === null) {
+			$sender->sendMessage(TextFormat::RED . $this->translateString("notinplot"));
+			return true;
+		}
+		if($plot->owner !== $sender->getName() and !$sender->hasPermission("myplot.admin.fill")) {
+			$sender->sendMessage(TextFormat::RED . $this->translateString("notowner"));
+			return true;
+		}
+		if(Item::fromString($args[0]) !== null && Item::fromString($args[0])->getBlock()->getId() !== Block::AIR) {
+            $maxBlocksPerTick = (int) $this->getPlugin()->getConfig()->get("FillBlocksPerTick",  256);
```
Comment (jasonwynn10) [inline src/MyPlot/subcommand/FillSubcommand.php:45]: seriously, whats with the partial indents?

File: plugin.yml [hunk 19]
```diff
@@ -128,4 +132,4 @@ permissions:
       myplot.claimplots.2:
         default: true
       myplot.claimplots.unlimited:
-        default: op
+        default: op
```
Comment (jasonwynn10) [inline plugin.yml:135]: line endings

**Review by jasonwynn10 — CHANGES_REQUESTED** [review]
Tons of formatting issues...

**Comment by Wertzui123** [comment]
I'm sorry for all the formatting issues, I think my IDE is configured to use spaces and myplot uses tabs.

File: resources/deu.ini (hunk 16 above)
Comment (Wertzui123) [reply resources/deu.ini:8]: Oops, you're right.

File: resources/deu.ini (hunk 15 above)
Comment (Wertzui123) [reply resources/deu.ini:295]: Oops, you're right.

File: src/MyPlot/MyPlot.php (hunk 3 above)
Comment (Wertzui123) [reply src/MyPlot/MyPlot.php:744]: It currently leaves the bedrock layer. I think replacing it isn't what most people want when using /p fill (e.g. with air).

File: src/MyPlot/MyPlot.php (hunk 2 above)
Comment (Wertzui123) [reply src/MyPlot/MyPlot.php:620]: This should be in the fillPlot function. I'll fix it.

File: src/MyPlot/subcommand/FillSubcommand.php (hunk 17 above)
Comment (Wertzui123) [reply src/MyPlot/subcommand/FillSubcommand.php:32]: What's the problem here?

File: plugin.yml (hunk 19 above)
Comment (Wertzui123) [reply plugin.yml:135]: Which line endings does MyPlot use?

File: src/MyPlot/MyPlot.php (hunk 8 above)
Comment (Wertzui123) [reply src/MyPlot/MyPlot.php:1207]: Your string concatenating is not consistent. Sometimes there are spaces and sometimes not.

File: resources/deu.ini (hunk 14 above)
Comment (Wertzui123) [reply resources/deu.ini:292]: Oops, you're right.

File: resources/deu.ini (hunk 13 above)
Comment (Wertzui123) [reply resources/deu.ini:281]: Oops, you're right.

File: src/MyPlot/forms/subforms/FillForm.php (hunk 11 above)
Comment (Wertzui123) [reply src/MyPlot/forms/subforms/FillForm.php:22]: I may add this when everything else works.

File: plugin.yml (hunk 19 above)
Comment (jasonwynn10) [reply plugin.yml:135]: I have my personal setup to view files in `CRLF` but git is set to only commit in `CR` endings

File: src/MyPlot/MyPlot.php [hunk 20]
```diff
@@ -686,6 +688,77 @@ public function clearPlot(Plot $plot, int $maxBlocksPerTick = 256) : bool {
 		return true;
 	}
 
+	/**
```
Comment (jasonwynn10) [inline src/MyPlot/MyPlot.php:691]: please use tabs instead of spaces

File: src/MyPlot/subcommand/FillSubcommand.php (hunk 17 above)
Comment (jasonwynn10) [reply src/MyPlot/subcommand/FillSubcommand.php:32]: spaces and its indented more than the lines below

**Review by jasonwynn10 — CHANGES_REQUESTED** [review]

File: plugin.yml (hunk 19 above)
Comment (jasonwynn10) [reply plugin.yml:135]: actually this line fixes a mistake so we can probably leave it

**Comment by Wertzui123** [comment]
I have fixed the remaining bugs and the plot filling feature now works as expected.
This pull request is now finally ready to be merged.

**Review by jasonwynn10 — APPROVED** [review]
